src/program_group.rs

```diff
@@ -0,0 +1,196 @@
+    /// to userspace. The priority number specified should be valid for the scheduling policy you
+    /// provide (documented in the enum). This may be useful if you find you're missing messages
+    /// that you expect to be present, or are dropping more messages than seems reasonable.
+    pub fn polling_thread_priority(mut self, scheduling_policy: SchedulingPolicy) -> Self {
+        self.polling_thread_policy = Some(scheduling_policy);
+        self
+    }
+
+    /// Attempt to load [`ProgramVersion`](struct@ProgramVersion)s until one
+    /// successfully loads.
+    ///
+    /// This function attempts to load each `ProgramVersion` in the order given until
+    /// one successfully loads. When one loads, if that version had a perfmap channel,
+    /// a [`PerfChannelMessage`](struct@PerfChannelMessage) receiver crossbeam channel
+    /// is available after loading by calling `get_receiver()` on the `ProgramGroup`.
+    /// If none load, a `NoProgramVersionLoaded` error is returned, along with all the
+    /// internal errors generated during attempted loading.
+    ///
+    /// If the program version contain any perfmaps,
+    /// perfmap_opts_fn(), will be called on each one until a version
+    /// suceeds to load. perfmap_opts_fn() returns a channel from
+    /// which to send perf messages, and a size (in bytes) for the per
+    /// cpu perf buffer. When the perfmap is created the buffer will
+    /// take at most the specified number of bytes but it will shrink
+    /// to fit a page size that is a multiple of two.
+    ///
+    /// NOTE: Once you call `load()`, it cannot be called again without re-creating
+    /// the `ProgramGroup`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use oxidebpf::ProgramBlueprint;
+    /// use oxidebpf::{ProgramGroup, Program, ProgramVersion, ProgramType};
+    /// use std::path::PathBuf;
+    ///
+    /// let program = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+    ///             .join("test")
+    ///             .join(format!("test_program_{}", std::env::consts::ARCH));
+    /// let program_blueprint =
+    ///     ProgramBlueprint::new(&std::fs::read(program).expect("Could not open file"), None)
+    ///         .expect("Could not open test object file");
+    /// let mut program_group = ProgramGroup::new();
+    ///
+    /// let (tx, rx) = crossbeam_channel::bounded(1024);
+    ///
+    /// program_group.load(
+    ///     program_blueprint,
+    ///     vec![ProgramVersion::new(vec![Program::new(
+    ///         "test_program",
+    ///         &["do_mount"],
+    ///     ).syscall(true)])],
+    ///     || (tx.clone(), 1024 * 8),
+    /// ).expect("Could not load programs");
+    /// ```
+    pub fn load(
+        &mut self,
+        program_blueprint: ProgramBlueprint,
+        program_versions: Vec<ProgramVersion<'a>>,
+        mut perfmap_opts_fn: impl FnMut() -> (Sender<PerfChannelMessage>, usize),
```

Comment:
My opinion is that this gives more control to the user:

1. The user can directly give us a channel, however they want to make it it's up to them. We no longer need to keep a "receiver" since we are just being given the sender. They could make an unbounded receiver and it would still work so we aren't boxing them in into one way of handling their events.

2. They can directly decide how they want to calculate the buffer size, so not necessarily channel_capacity * elems. If they believe their cpu can keep up then they can have less of a perf buffer.